jvm-runtime/ftl-runtime/common/runtime/src/main/java/xyz/block/ftl/runtime/FTLController.java

```diff
@@ -234,4 +222,63 @@ private ModuleContextResponse getModuleContext() {
         }
     }
 
+    private class ModuleObserver implements StreamObserver<ModuleContextResponse> {
+
+        final AtomicInteger failCount = new AtomicInteger();
+
+        @Override
+        public void onNext(ModuleContextResponse moduleContextResponse) {
+            synchronized (this) {
+                currentError = null;
+                FTLController.this.moduleContextResponse = moduleContextResponse;
+                if (waiters) {
+                    this.notifyAll();
+                    waiters = false;
+                }
+            }
+
+        }
+
+        @Override
+        public void onError(Throwable throwable) {
+            log.error("GRPC connection error", throwable);
+            synchronized (this) {
+                currentError = throwable;
+                if (waiters) {
+                    this.notifyAll();
+                    waiters = false;
+                }
+            }
+            if (failCount.incrementAndGet() < 5) {
+                verbService.getModuleContext(ModuleContextRequest.newBuilder().setModule(moduleName).build(), moduleObserver);
+            }
+        }
+
+        @Override
+        public void onCompleted() {
+            onError(new RuntimeException("connection closed"));
+        }
+    }
+
+    public record Datasource(String connectionString, String username, String password) {
+
+        public static Datasource fromDSN(String dsn) {
+            //TODO: this is horrible, just quick hack for now
```

Comment:
😂

Reply:
Lots of cleanup is needed, this was mostly about making sure that I could integrate everything the way I expected to. I should have a cleaned up version at some point on Monday.